### ISSUE TITLE
Make translations importable at build-time?

### ISSUE THREAD

**Issue by hackel** [andywer/laravel-js-localization#21]
I don't generate my frontend javascript html pages with PHP/Laravel, so I can't include the templates to pull translations as additional http requests.  I could still include them manually, in theory, but ideally, I would like to import them directly into my application at build-time, with gulp (and browserify in my case).  It looks like this is already 90% of the way there, I think it just needs a command to export the cached translation files to stdout instead of an http response, and then a quick gulp wrapper could be written, or even better, an extension to Laravel Elixir.

Also, everything should be exportable as an es6 module so that it doesn't pollute the global/window scope!

**Comment by andywer** [comment]
Hey there! Have a look at the develop branch (composer dependency: `"andywer/js-localization": "dev-develop"`). Does that command work for you?

**Comment by andywer** [comment]
An idea: The export command should probably refresh the translations cache automatically.

**Comment by hackel** [comment]
Hey, thanks for working on this.  It is exporting my translations just fine, but I'm not able to do anything with the generated file itself, since it still relies on a global Lang object.  I have figured out a way to solve this, though (I think!).  Basically, in the generated file, I first need to import the localization script into the Lang object and then return it.  e.g.:

``` js
import Lang from '../../vendor/andywer/js-localization/resources/js/localization';
Lang.addMessages(...);
export default Lang;
```

This step could be eliminated if the export were to contain just a POJO, which I could then run through Lang.addMessages() myself.  I think this way makes the most sense, but requires an extra step to maintain backward compatibility.

I also had to convert the localization script to a [UMD](https://github.com/umdjs/umd) module.  I will submit a PR for this change to see what you think.  I didn't test loading the localization.js file from html, but I believe it should still work as it does now.

With these changes, I was easily able to return Lang as an Angular service and use it throughout my app!

I definitely agree the export command should refresh the translations cache automatically, or really just bypass it entirely.  A config option will also need to be added to prevent loading the routes file.

**Comment by bytestream** [comment]
The idea behind this is great. I'm hoping you're willing to go one step further:

I don't see the purpose of serving the JavaScript file via PHP - it adds a lot of unnecessary overhead. Is there any reason why you can't take the export command one step further?

``` php
class ExportJavaScriptLanguages extends Command
{
    /**
     * The console command name.
     *
     * @var string
     */
    protected $signature = 'javascript-localisation:export {outputFile : Where to write the minified file to.}';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = "Export translations as Javascript file";

    /**
     * Execute the console command.
     *
     * @return void
     */
    public function fire()
    {
        // Get the output file
        $outputFile = $this->argument('outputFile');

        // Clear the cache
        Artisan::call('js-localization:refresh');

        // Get the JavaScript data
        $contents = file_get_contents(base_path('vendor/andywer/js-localization/public/js/localization.min.js'));
        $contents .= "\n";
        $contents .= $this->getMessagesJson();
        $contents .= $this->getConfigJson();

        // Write the file
        if (! file_exists(dirname($outputFile))) {
            mkdir(dirname($outputFile), 0777, true);
        }
        File::put($outputFile, $contents);
        $this->line('Wrote Javascript file: '.$outputFile);
    }

    /**
     * Transforms the cached data to stay compatible to old versions of the package.
     *
     * @param string $messages
     * @return string
     */
    protected function ensureBackwardsCompatibility($messages)
    {
        if (preg_match('/^\\{"[a-z]{2}":/', $messages)) {
            return $messages;
        } else {
            return '{"' . app()->getLocale() . '":' . $messages . '}';
        }
    }

    /**
     * Get the configured messages from the translation files
     *
     * @return string
     */
    protected function getMessagesJson()
    {
        $messages = MessageCachingService::getMessagesJson();
        $messages = $this->ensureBackwardsCompatibility($messages);
        $contents  = 'Lang.addMessages(' . $messages . ');';
        return $contents;
    }

    /**
     * Get the JSON-encoded config properties that shall be passed to the client.
     *
     * @return string
     */
    protected function getConfigJson()
    {
        $config = ConfigCachingService::getConfigJson();
        $contents = 'Config.addConfig(' . $config . ');';
        return $contents;
    }
}
```

We can now add the command to our build system (gulp or grunt), have it watch the language files for changes, and automatically update the JavaScript file. For example:

```
gulp.task('refresh-js-langs', shell.task('php artisan javascript-localisation:export resources/assets/lang/lang.min.js'));

gulp.task('watch', function() {
    watch('./resources/lang/**/*.php', { verbose: true }, function () {
        gulp.run(['refresh-js-langs']);
    });
});
```

This cuts the overhead down on the `js-localisation/all.js` route from ~900ms to ~150ms as we can instead access `resources/assets/lang/lang.min.js` directly.

**Comment by tobia** [comment]
I agree with @bytestream 

I'm already using Elixir, with Browserify and Version plugins, to generate a single js file at deploy time, and `gulp watch` to keep it updated while I'm developing.

I think it makes sense to integrate the translations into this flow, which I believe is Laravel's recommended setup, instead of serving files via PHP.

**Comment by andywer** [comment]
It's all very true. I just lack the time and (to be honest) the motivation to take this one step further.

I will review your PRs, though, if you want to implement it. Maybe there is someone out there willing and able to maintain the project in the long-term.

**Comment by Jarlskov** [comment]
This would be fixed by PR #35 , wouldn't it?

**Comment by andywer** [comment]
@Jarlskov Good point! I hope it solves this issue :)
